### PR TITLE
To element reduction of public inputs and proof context

### PR DESCRIPTION
**Pull request by irakliyk**
This PR implements reduction to elements of public inputs and proof context. This will allow `RandomCoin` to be initialized from a list of field elements (rather than bytes). The advantage of this approach is that it simplifies handling of random coin in the context of recursive proofs.

### CODE REVIEW

**Comment by irakliyk** [comment]
@Nashtare, @Al-Kindi-0 - would appreciate a review on this.

File: air/src/options.rs [hunk 1]
```diff
@@ -206,6 +208,22 @@ impl ProofOptions {
     }
 }
 
+impl<E: StarkField> ToElements<E> for ProofOptions {
+    fn to_elements(&self) -> Vec<E> {
+        // encode filed extension and FRI parameters into a single field element
```
Comment (Al-Kindi-0) [inline air/src/options.rs:213]: nit: field

File: air/src/proof/context.rs [hunk 2]
```diff
@@ -93,6 +93,45 @@ impl Context {
...
+        let (m1, m2) = self.field_modulus_bytes.split_at(num_modulus_bytes / 2);
+        result.push(bytes_to_element(m1));
+        result.push(bytes_to_element(m2));
+
+        // convert proof options and trace length to elements
+        result.append(&mut self.options.to_elements());
+        result.push(E::from(self.trace_length as u64));
+
+        // convert trace metadata to elements; this is done by breaking trace metadata into chunks
+        // of bytes which are slightly smaller than field element requires, and then converting
```
Comment (Al-Kindi-0) [inline air/src/proof/context.rs:123]: nit: sentence containing "requires" should probably be reformulated

Comment (Al-Kindi-0) [reply air/src/proof/context.rs:123]: This sentence from the same PR should work: "..is slightly smaller than the number of bytes needed to encode a field element"

**Review by Al-Kindi-0 — COMMENTED** [review]
Looks great, thank you!
I left only some small nits.

**Comment by Nashtare** [comment]
Nothing else to add 🙂

File: air/src/proof/context.rs (hunk 2 above)
Comment (irakliyk) [reply air/src/proof/context.rs:123]: Thank you! Fixed!

**Pull request merged**